docs: add sphinx documentation

jia200x opened the pull request:
This PR adds initial support for Sphinx documentation.

I suggest to read [this](https://brendanhasz.github.io/2019/01/05/sphinx.html) to have a better understanding of the usage The root file Sphinx file is the `index.rst`, from there we can include other files and module documentation.

To build the documentation:
- `pipenv run doc`

This command will call Sphinx's `make html` command to build HTML documentation. Other documentation formats are also available (man pages, latex, JSON, XML, etc).

The documentation is placed under `_build/html/index.html`

This PR also adds CircleCI build tests for the documentation, so we ensure the documentation is always consistent.

We need to slowly migrate all functions to their Sphinx format.

leandrolanzieri commented:
> We need to slowly migrate all functions to their Sphinx format.

I suggest we make either a tracking issue or a project for this migration.

sasalatart commented on Makefile:
```diff
@@ -0,0 +1,20 @@
+# Minimal makefile for Sphinx documentation
```
I am not very familiar with `Makefile`s. Is there any possibility of placing doc-related things inside a `docs` dir or something like that?

Having a `Makefile` in the root dir of the repo is a little bit strange. Someone from the outside seeing the Makefile in the root dir might think that the whole app is built using Make. If the answer is no, then it's no problem.

sasalatart replied:
And then somehow calling `make` from the `Pipfile`, but pointing to `docs/Makefile`

sasalatart replied:
(This is not a necessary change though, just a question)

leandrolanzieri replied:
I'm thinking if having this Makefile is actually really necessary. Perhaps we can just keep the commands on pipenv (as far as we can).

sasalatart replied:
Yeah, if that approach is possible I would vote for that!

jia200x replied:
We can get rid of the Makefile. I can do it :)

felipehuerta17 commented:
Hi @jia200x , I pulled your branch and tried

`pipenv run doc`

But I got the following error message:

![image](https://user-images.githubusercontent.com/33637198/78609531-c1152680-785a-11ea-81c3-82e480035edf.png)

Do you know what may be the reason? Should everyone receive that Theme error?

felipehuerta17 commented:
@sasalatart likes Sphinx because [this extension](https://brendanhasz.github.io/2019/01/05/sphinx.html#napoleon-extension)

jia200x commented:
> @sasalatart likes Sphinx because this extension

Actually we should indeed use that one!

jia200x commented:
@sasalatart it seems the CircleCI script is failing. Have you had an issue like that before?

jia200x commented:
Just to save some time, I simply migrated the whole API to Sphinx.

The format of the "index.rst" is not the best, but it's a base for follow up work.
I also addressed the `pipenv run doc` request so now pipenv directly invokes the `sphinx-build` command. I removed the Makefile and make.bat files.

jia200x commented:
let's wait for #26

felipehuerta17 commented:
@jia200x will rebasing this to master solve the ci/circleci complain?
